packages/smart-router/evm/trade.ts

```diff
@@ -0,0 +1,73 @@
+import invariant from 'tiny-invariant'
+import {
+  Currency,
+  TradeType,
+  Percent,
+  ZERO,
+  CurrencyAmount,
+  Fraction,
+  ONE,
+  Price,
+  computePriceImpact,
+} from '@pancakeswap/sdk'
+
+import { TradeWithStableSwap } from './types'
+import { getMidPrice } from './utils/route'
+
+export const Trade = {
+  maximumAmountIn,
+  minimumAmountOut,
+  executionPrice,
+  priceImpact,
+}
+
+/**
+ * Get the maximum amount in that can be spent via this trade for the given slippage tolerance
+ */
+function maximumAmountIn<TIn extends Currency, TOut extends Currency, TTradeType extends TradeType>(
+  trade: TradeWithStableSwap<TIn, TOut, TTradeType>,
+  slippageTolerance: Percent,
+): CurrencyAmount<TIn> {
+  invariant(!slippageTolerance.lessThan(ZERO), 'SLIPPAGE_TOLERANCE')
+  if (trade.tradeType === TradeType.EXACT_INPUT) {
```

Comment:
is this condition always be true while we are using smart router?

Reply:
since we now support both exact in and exact out, so we could still have different conditions met. Like when the trade is v2 not mixed